proposals/NNNN-task-executor-preference.md

```diff
@@ -0,0 +1,611 @@
+
+Notably, the task executor preference is in effect for the entire structured task hierarchy while running in a task or scope where a task executor preference is set. For example, the following snippet illustrates child tasks created inside of a `withTaskExecutor`.
+
+If a task with such executor preference encounters code which is `isolated` to some specific actor, it would adhere to that requirement and hop to it as expected:
+
+```swift
+let capy: Capybara = Capybara()
+actor Capybara { func eat() {} } 
+
+Task(on: executor) {
+  // starts on 'executor', however...
+  try await capy.eat() // still executes actor isolated code on the actor's executor, as expected
+}
+```
+
+## Detailed design
+
+### Setting task executor preference
+
+A new concept of task executor preference is added to Swift Concurrency tasks. This preference is stored in a task and propagated throughout child tasks (such as ones created by TaskGroups and async let).
+
+The preference can be set using various APIs that will be discussed in detail in their respective sections. The first of those APIs is `withTaskExecutor` which is can be called inside an asynchronous context to both ensure we’re executing on the expected executor, as well as set the task executor preference for the duration of the operation closure:
+
+```swift
+await withTaskExecutor(someExecutor) { 
+ // guaranteed to be executing on someExecutor
+}
+```
+
+Once set, the effect of an executor preference is such that a nonisolated func instead of immediately hopping to the global pool, it may hop to the preferred executor, e.g.:
+
+```swift
+nonisolated func tryMe() async { 
+ // ...
+}
+
+let preferredExecutor: SomeConcreteTaskExecutor = ...
+Task(on: preferredExecutor) {
+  // executes on 'preferredExecutor'
+  await tryMe() // tryMe body would execute on 'preferredExecutor'
+}
+
+ await tryMe() // tryMe body would execute on 'default global concurrent executor'
+```
+
+### The `TaskExecutor` protocol
+
+In order to fulfil the requirement that we'd like default actors to run on a task executor, if it was set, we need to introduce a new kind of executor.
+
+This stems from the fact that `SerialExecutor` and how a default actor effectively acts as an executor "for itself" function in Swift.
+A default actor (so an actor which does not use a custom executor), has a "default" executor that is created by the Swift runtime and uses the actor is the executor's identity.
+This means that the runtime executor tracking necessarily needs to track that some code is executing on a specific serial executor in order for things like `assumeIsolated` or the built-in runtime thread-safety checks can utilize them.
+
+The new protocol mirrors `Executor` and `SerialExecutor` in API, however it provides different semantics, and is tracked using a different mechanism at runtime -- by obtaining it from a task's executor preference record.
+
+The `TaskExecutor` is defined as:
+
+```swift
+public protocol TaskExecutor: Executor {
```

Comment:
I think we should provide some docs here on what is allowed. A `TaskExecutor` is allowed to be a pool of threads or a single thread. Both of them are fine and we should call it out on the docs.